frontend/apps/web/components/RunTImeline/RunTimeline.tsx

```diff
@@ -0,0 +1,457 @@
+import {
+  DropdownMenu,
+  DropdownMenuCheckboxItem,
+  DropdownMenuContent,
+  DropdownMenuTrigger,
+} from '@/components/ui/dropdown-menu';
+import {
+  Tooltip,
+  TooltipContent,
+  TooltipProvider,
+  TooltipTrigger,
+} from '@/components/ui/tooltip';
+import { cn } from '@/libs/utils';
+import { Timestamp } from '@bufbuild/protobuf';
+import { JobRunEvent, JobRunStatus } from '@neosync/sdk';
+
+import { JobRunStatus as JobRunStatusEnum } from '@neosync/sdk';
+import {
+  CheckCircledIcon,
+  CrossCircledIcon,
+  MinusCircledIcon,
+  MixerHorizontalIcon,
+} from '@radix-ui/react-icons';
+import {
+  addMilliseconds,
+  format,
+  formatDuration,
+  intervalToDuration,
+} from 'date-fns';
+import { ReactElement, useMemo, useState } from 'react';
+import Spinner from '../Spinner';
+import TruncatedText from '../TruncatedText';
+import { Badge } from '../ui/badge';
+import { Button } from '../ui/button';
+
+interface Props {
+  tasks: JobRunEvent[];
+  jobStatus?: JobRunStatusEnum;
+}
+
+export default function RunTimeline(props: Props): ReactElement {
+  const { tasks, jobStatus } = props;
+
+  // this should probably be better typed and using the ActivityStatus types
+  // but those aren't currenlty sent to the FE as a status but are represented in the
+  // type field
+  const [selectedStatuses, setSelectedStatuses] = useState<string[]>([
+    'running',
+    'completed',
+    'failed',
+    'canceled',
+  ]);
+
+  const formatFullDate = (date: Timestamp | Date | undefined) => {
+    if (!date) return 'N/A';
+
+    if (date instanceof Timestamp) {
+      return format(convertTimestampToDate(date), 'MM/dd/yyyy HH:mm:ss:SSS');
+    }
+
+    if (date instanceof Date) {
+      return format(date, 'MM/dd/yyyy HH:mm:ss:SSS');
+    }
+  };
+
+  const formatDate = (date: Date) => format(date, 'MM/dd/yyyy');
+  const formatTime = (date: Date) => format(date, 'HH:mm:ss:SSS');
+
+  const formatTaskDuration = (
+    s: Timestamp | undefined,
+    end: Date | undefined
+  ) => {
+    if (!s || !end) return 'N/A';
+    const start = convertTimestampToDate(s);
+    const duration = intervalToDuration({ start, end });
+    const milliseconds = end.getTime() - start.getTime();
+    const millis = milliseconds % 1000;
+
+    // format the duration string
+    const formattedDuration = formatDuration(duration, {
+      format: ['hours', 'minutes', 'seconds'],
+      delimiter: ', ',
+    });
+
+    if (!formattedDuration) {
+      return `${millis} ms`;
+    }
+
+    return `${formattedDuration}${millis > 0 ? `, ${millis} ms` : ''}`;
+  };
+
+  const { timelineStart, totalDuration, timeLabels } = useMemo(() => {
+    // find earliest start date out of all of the activities
+    let startTime = Infinity;
+    let endTime = -Infinity;
+
+    tasks.map((t) => {
```

Comment:
it's a lot clearer here if you do tasks.forEach instead.
map is generally used when you are returning stuff into a new list.

Reply:
updated